integration/ruler_test.go

```diff
@@ -179,45 +182,106 @@ func TestRulerEvaluationDelay(t *testing.T) {
+	require.NoError(t, writeFileToSharedDir(s, filepath.Join("ruler_configs", user, namespace), []byte(cortexRulerEvalStaleNanConfigYaml)))
 	cortex := e2ecortex.NewSingleBinaryWithConfigFile("cortex", cortexConfigFile, configOverrides, "", 9009, 9095)
 	require.NoError(t, s.StartAndWaitReady(cortex))
 
 	// Create a client with the ruler address configured
-	c, err := e2ecortex.NewClient("", cortex.HTTPEndpoint(), "", cortex.HTTPEndpoint(), "")
+	c, err := e2ecortex.NewClient(cortex.HTTPEndpoint(), cortex.HTTPEndpoint(), "", cortex.HTTPEndpoint(), "")
 	require.NoError(t, err)
 
-	// Wait until the rule is evaluated
-	require.NoError(t, cortex.WaitSumMetrics(e2e.Greater(0), "cortex_prometheus_rule_evaluations_total"))
-
 	now := time.Now()
 
-	result, err := c.QueryRange("time_eval", now.Add(-10*time.Minute), now, time.Minute)
+	// Generate series that includes stale nans
+	var samplesToSend int = 10
+	series := prompb.TimeSeries{
+		Labels: []prompb.Label{
+			{Name: "__name__", Value: "a_sometimes_stale_nan_series"},
+			{Name: "instance", Value: "sometimes-stale"},
+		},
+	}
+	series.Samples = make([]prompb.Sample, samplesToSend)
+	posStale := map[int]struct{}{
+		2: {},
+	}
+
+	// Create samples, that are delayed by the evaluation delay with increasing values.
+	for pos := range series.Samples {
+		series.Samples[pos].Timestamp = e2e.TimeToMilliseconds(now.Add(-evaluationDelay)) + 1000*int64(pos)
+		series.Samples[pos].Value = float64(pos + 1)
+
+		// insert staleness marker at the positions marked by posStale
+		if _, stale := posStale[pos]; stale {
+			series.Samples[pos].Value = math.Float64frombits(value.StaleNaN)
+		}
+	}
+
+	// Insert metrics
+	res, err := c.Push([]prompb.TimeSeries{series})
 	require.NoError(t, err)
-	require.Equal(t, model.ValMatrix, result.Type())
+	require.Equal(t, 200, res.StatusCode)
 
-	// Iterate through the values recorded and ensure they exist in the past.
-	matrix := result.(model.Matrix)
+	// Wait until the rule is evaluated for the first time
+	require.NoError(t, cortex.WaitSumMetrics(e2e.Greater(1.0), "cortex_prometheus_rule_evaluations_total"))
+
+	// Query the timestamp of the latest result to ensure the evaluation is delayed
+	result, err := c.Query("timestamp(stale_nan_eval)", now)
+	require.NoError(t, err)
+	require.Equal(t, model.ValVector, result.Type())
+
+	vector := result.(model.Vector)
+	require.Equal(t, 1, vector.Len(), "expect one sample returned")
 
 	// 290 seconds gives 10 seconds of slack between the rule evaluation and the query
-	// to account for CI latency, but ensures the latest evalation was in the past.
-	var maxDiff int64 = 290
+	// to account for CI latency, but ensures the latest evaluation was in the past.
+	var maxDiff int64 = 290_000
+	require.GreaterOrEqual(t, e2e.TimeToMilliseconds(time.Now())-int64(vector[0].Value)*1000, maxDiff)
+
+	// Wait until the rule is evaluated once per sample sent
```

Comment:
I'm not sure I do understand this. Why you need to wait `samplesToSend` evaluations?

Reply:
I have added a comment to clarify:
```
	// Wait until all the pushed samples have been evaluated by the rule. This
	// ensures that rule results are successfully written even after a
	// staleness period.```